mmengine/model/wrappers/pipeline_distributed.py

```diff
@@ -0,0 +1,1213 @@
+                            cuda_device = cuda_devices[cuda_pointer]
+                            modules[module_pointer]['exec'] = cuda_device
+                            if module_pointer != self.num_pipelines - 1:
+                                module_pointer += 1
+                                cuda_pointer += 1
+                                cuda_pointer %= len(cuda_devices)
+                                meta_info['module_pointer'] = module_pointer
+                                meta_info['cuda_pointer'] = cuda_pointer
+                else:
+                    if module_pointer == self.num_pipelines - 1:
+                        raise RuntimeError(
+                            'The model is too large to fit into' +
+                            f'{cuda_devices[cuda_pointer]}' +
+                            'Please use more GPUs.')
+                # if the current module can not be put into the current device
+                # but we can go to the next module
+                # we can put the current module into the next device
+                if module_pointer != self.num_pipelines - 1:
+                    cuda_device = cuda_devices[cuda_pointer]
+                    modules[module_pointer]['exec'] = cuda_device
+                    module_pointer += 1
+                    cuda_pointer += 1
+                    cuda_pointer %= len(cuda_devices)
+                    meta_info['module_pointer'] = module_pointer
+                    meta_info['cuda_pointer'] = cuda_pointer
+                    modules[module_pointer]['modules'].append(name)
+                    modules[module_pointer]['flops'] += module_flops
+                    modules[module_pointer]['parameter_size'] += module_size
+
+        dfs(self.model_tree, name='', meta_info=meta_info)
+        # handle last
+        if modules[-1]['exec'] is None:
+            modules[-1]['exec'] = cuda_devices[meta_info['cuda_pointer']]
+        # infer init
+        cpu_used = 0
+        for i in range(self.num_pipelines):
+            if i < len(cuda_devices):
+                modules[i]['init'] = cuda_devices[i]
+            else:
+                cpu_used += modules[i]['parameter_size']
+                if cpu_used > self.memory_map['cpu']:
+                    # if the cpu memory is not enough
+                    # we can use disk offload
+                    modules[i]['init'] = 'disk'
+                else:
+                    modules[i]['init'] = 'cpu'
+
+        # format modules
+        device_map = {}
+        for i in range(self.num_pipelines):
+            modules_i = modules[i]
+            for module in modules_i['modules']:
+                device_map[module] = {
+                    'part_id': i,
+                    'init_device': modules_i['init'],
+                    'exec_device': modules_i['exec']
+                }
+        # handle tied weights
+        tied_weights = self.model_tree['tied_parameters']
+        for source, targets in tied_weights.items():
```

Comment:
The key of tied_weights means `param_name`, and value means list of `module_name`s. So, why do we use `device_map[source]` and `device_map[target]` here?